File: tools/reference/node/shape-arc-pie.mjs
// FIX-S04: independent d3-shape 3.2.0 arc geometry, centroids and pie ordering/layout.
import * as d3 from 'd3-shape';
import fs from 'node:fs';
import path from 'node:path';
import {root,provenance,writeJson,digest} from './provenance.mjs';
const out=path.resolve(process.argv[2]||path.join(root,'fixtures/shapes'));
class Context {
 constructor(){this.operations=[];}
 moveTo(...args){this.operations.push({op:'moveTo',args});}
 lineTo(...args){this.operations.push({op:'lineTo',args});}
 arc(x,y,r,a0,a1,anticlockwise=false){this.operations.push({op:'arc',args:{x,y,r,a0,a1,anticlockwise}});}
 closePath(){this.operations.push({op:'closePath'});}
}
const arcs=[];
function add(id,input,config={}){
 const g=d3.arc();for(const [key,value]of Object.entries(config))g[key](value);
 const context=new Context();let error=null;
 try{g.context(context)(input);}catch(e){error=String(e);}
 const centroid=g.centroid(input),svg={};
 for(const digits of [0,3,12,null]){try{svg[String(digits)]=g.context(null).digits(digits)(input);}catch(e){svg[String(digits)]={error:String(e)};}}
 const finite=context.operations.every(o=>!o.args||Object.values(o.args).every(v=>typeof v!=='number'||Number.isFinite(v)))&&centroid.every(Number.isFinite);
 arcs.push({id,input,config,operations:context.operations,centroid,svg,finite,error});
}
for(const [ri,ro] of [[0,10],[3,10],[9,10],[10,4],[-3,10],[-3,-1],[0,0]])for(const span of [0,1e-14,1e-8,.03,.5,Math.PI,Math.PI*1.7,Math.PI*2,Math.PI*3,-1.2])for(const corner of [0,.5,9]){
 add(`r${ri}-${ro}-a${span}-c${corner}`,{innerRadius:ri,outerRadius:ro,startAngle:.3,endAngle:.3+span,padAngle:.12},{cornerRadius:corner});
}
for(const start of [0,-Math.PI,7,1e6])for(const span of [.001,.2,1,-1,Math.PI*2])for(const pad of [0,.001,.4,2,-.1])for(const padRadius of [null,0,2,20]){
 add(`padding-${arcs.length}`,{innerRadius:5,outerRadius:10,startAngle:start,endAngle:start+span,padAngle:pad},{cornerRadius:2,padRadius});
}
for(const overrides of [{innerRadius:7},{outerRadius:2},{startAngle:-1},{endAngle:4},{padAngle:.7},{innerRadius:0,outerRadius:12,startAngle:0,endAngle:Math.PI/2,cornerRadius:2,padRadius:16}])add(`constant-${arcs.length}`,{innerRadius:3,outerRadius:10,startAngle:1,endAngle:2,padAngle:.1},overrides);
for(const input of [{innerRadius:0,outerRadius:10,startAngle:0,endAngle:Math.PI/2},{innerRadius:1e-10,outerRadius:1e-9,startAngle:0,endAngle:1},{innerRadius:1e100,outerRadius:2e100,startAngle:0,endAngle:1},{innerRadius:1e300,outerRadius:2e300,startAngle:0,endAngle:1}])add(`boundary-${arcs.length}`,input,{cornerRadius:1});
const pies=[];
for(const values of [[],[1,1,2],[2,-1,0,3],[0,0,0],[-1,-2],[1,1,1],[1e-300,2e-300],[1e300,2e300]])for(const order of ['default','none','valuesAscending','dataDescending'])for(const [start,end,pad]of [[0,Math.PI*2,0],[0,Math.PI*2,.2],[1,-2,.1],[0,100,10],[2,2,.2],[0,1,-.1]]){
 const data=values.map((value,i)=>({id:String(9007199254741001n+BigInt(i)),label:String.fromCharCode(90-i),value})),g=d3.pie().value(d=>d.value).startAngle(start).endAngle(end).padAngle(pad);
 if(order==='none')g.sort(null);if(order==='valuesAscending')g.sortValues((a,b)=>a-b);if(order==='dataDescending')g.sort((a,b)=>b.label.localeCompare(a.label,'en'));
 pies.push({id:`pie-${pies.length}`,data,order,start,end,pad,result:g(data)});
}
writeJson(path.join(out,'arc-pie.json'),JSON.parse(JSON.stringify({schema_version:1,arcs,pies},(_,v)=>typeof v==='number'&&!Number.isFinite(v)?{number:String(v)}:v)));
writeJson(path.join(out,'arc-pie-manifest.json'),{schema_version:1,references:['d3-shape','d3-path'].map(provenance),arc_cases:arcs.length,pie_cases:pies.length,generator_sha256:digest(fs.readFileSync(new URL(import.meta.url))),cases_sha256:digest(fs.readFileSync(path.join(out,'arc-pie.json')))});
console.log('PASS independent arc/pie oracle:',arcs.length,'arcs and',pies.length,'pie layouts; implementation acceptance is separate.');

File: scripts/bindings/shape_arc_pie.py
"""FIX-S04 actual Python arc/pie constants, ownership and materialized source identity."""
from pathlib import Path
import sys,json,math,re
ROOT=Path(__file__).resolve().parents[2]
sys.path[:0]=[str(ROOT/'packages/python'),str(Path(sys.argv[1]).resolve())]
import finstack_chart as c
corpus=json.loads((ROOT/'fixtures/shapes/arc-pie.json').read_text())
def snake(k):return re.sub(r'(?<!^)(?=[A-Z])','_',k).lower()
def near(a,b):
    if isinstance(a,(int,float)) and isinstance(b,(int,float)):assert abs(a-b)<=2e-12*max(1,abs(b)),(a,b)
    elif isinstance(a,list):
        assert len(a)==len(b)
        for x,y in zip(a,b):near(x,y)
    elif isinstance(a,dict):
        assert a.keys()==b.keys()
        for k in a:near(a[k],b[k])
    else:assert a==b,(a,b)
negative=0
for case in corpus['arcs']:
    config={snake(k):v for k,v in case['config'].items()};datum={snake(k):v for k,v in case['input'].items()}
    arc=c.ShapeArc(config);copy=arc.copy();assert arc.config()==copy.config();arc.dispose()
    if not case['finite']:
        try:copy.generate(datum);assert False,case['id']
        except c.ChartError as error:assert error.code=='CHART_NUMERICAL_DOMAIN'
        copy.dispose();negative+=1;continue
    path=copy.generate(datum);retained=path.copy();expected=c.Path(3).apply_batch(case['operations'])
    near(path.result()['geometry']['commands'],expected.result()['geometry']['commands']);near(copy.centroid(datum),case['centroid']);assert path.to_svg()==case['svg']['3'],case['id']
    again=copy.generate(datum);assert again.result()==path.result();again.dispose();copy.dispose();original=retained.result();path.move_to(998,997);path.dispose();assert retained.result()==original;retained.dispose();expected.dispose()
count=0
for case in corpus['pies']:
    if case['order']=='dataDescending':continue # native comparator fixture; portable custom protocols belong to WP-S07.
    pie=c.ShapePie({'order':{'default':'ValuesDescending','none':'Input','valuesAscending':'ValuesAscending'}[case['order']],'angles':{'start_angle':case['start'],'end_angle':case['end'],'pad_angle':case['pad']}})
    owner=pie.copy();pie.dispose();data=case['data'];values=[d['value'] for d in data];result=owner.layout(data,values);again=owner.layout(data,values);owner.dispose();assert result==again
    expected=[{snake(k):v for k,v in arc.items()} for arc in case['result']];near(result,expected)
    if data:assert result[0]['data']['id']==data[0]['id'];result[0]['data']['label']='mutated';assert again[0]['data']==data[0]
    count+=1
assert count==144 and negative==1
constant=c.ShapeArc({'inner_radius':0,'outer_radius':10,'start_angle':0,'end_angle':math.pi/2});p=constant.generate();assert p.to_svg()=='M0,-10A10,10,0,0,1,10,0L0,0Z';p.dispose();constant.dispose()
assert c.ShapePie({'value':2}).layout([4,7])[0]['value']==2
for Type,config in [(c.ShapeArc,{'wrong':1}),(c.ShapePie,{'order':'Unknown'})]:
    try:Type(config);assert False
    except c.ChartError:pass
for run in [lambda:c.ShapeArc().generate(),lambda:c.ShapePie().layout(['a'],[]),lambda:c.ShapePie({'limits':{'max_points':0}}).layout([1]),lambda:c.ShapeArc({'limits':{'max_points':0}}).centroid()]:
    try:run();assert False
    except c.ChartError:pass
print('PASS Python arc/pie: 620 arc cases (619 finite, one checked overflow), 144 portable pie layouts, exact metadata, constants, copies, disposal and independent paths. 48 native comparator fixtures remain Rust-only until WP-S07.')
